### DISCUSSION TITLE
Uploading and downloading files as base64 string

### DISCUSSION THREAD

**Discussion by VaZark** [directus/directus#11753] [Feature Requests]
Can base64 encoding be implemented as an additional standard to upload and download files? File metadata has to be passed as a separate field though.

**Comment by joselcvarela** [comment]
Hey,
Can you tell your use case? This seems to be an extremely edge case 😅

**Reply by VaZark** [reply]
So the use case is targeted towards Serverless environments that have official support only for JSON format. 

In my case, when integrating with AWS Lambda + API Gateway, it is far easier to send/receive files sent as base64 encoded strings as support for forms are not natively supported on the official frameworks such as [AWS Chalice](https://aws.github.io/chalice/index.html). While it does allow reading raw binary data, that requires creating a form parser for every separate project. Base64 is a decent compromise when we know files will be relatively small

**Comment by rijkvanzanten** [comment]
> Base64 is a decent compromise when we know files will be relatively small

This is an important caveat. A base64 encoded file is on average about 30% larger than the original file, and seeing it's all plain text in JSON, you have the risk of running into memory problems, as the whole object is read into memory before being saved.

**Comment by rijkvanzanten** [comment]
One nice bonus of this is that you could theoretically upload the file relationally, so instead of using the regular `/files/upload` endpoint, you could associate the file directly when creating another item:

```
POST /items/articles

{
  "title": "Hello World",
  "featured_image": "data:image/jpeg;base64,/9j/2wCEAAgGBgcGBQgHBy......"
}
```

**Reply by NilsBaumgartner1994** [reply]
I support this. In my application i dont want to allow everyone to upload files. I want to give permission to only upload a specific file/field. Therefore a usecase for the js-sdk where i can upload a base64 for a field, which then gets transformed to an image would be realy nice.

**Comment by NilsBaumgartner1994** [comment]
Implemented it myself with an hook. Needs further refactoring but i will share my code here and update my solution later:

```
const { toArray } = require('@directus/shared/utils');

// Source: http://stackoverflow.com/questions/20267939/nodejs-write-base64-image-file
function decodeBase64Image(dataString) 
        {
          var matches = dataString.match(/^data:([A-Za-z-+\/]+);base64,(.+)$/);
          var response = {};

          if (matches.length !== 3) 
          {
            return new Error('Invalid input string');
          }

          let extension = dataString.substring(dataString.indexOf('/') + 1, dataString.indexOf(';base64'));

          response.type = matches[1];
          response.data = new Buffer.from(matches[2], 'base64');
          response.extension = extension;

          return response;
}

//in the filterAction for a specific collection
async function base64Hook(collection_name, file_field_name, payload, input, database, schema, accountability, registerFunctions, context){
        let storage = toArray(context.env.STORAGE_LOCATIONS)[0];

        let services = context.services;
        const {FilesService, FieldsService} = services;

        const adminAccountAbility = JSON.parse(JSON.stringify(accountability)); //make a copy !
        adminAccountAbility.admin = true; //usefull if we realy want to upload something as admin

        let filesService = new FilesService({schema, accountability: adminAccountAbility});
        let fieldsService = new FieldsService({schema, accountability: adminAccountAbility});

        let body = {folder: null, type: "image/png", storage: storage, filename_download: "test.png"}; //TODO get the correct folder
        let image = JSON.parse(JSON.stringify(payload.image));
        if(!!image){
             payload.image = null;
                if(typeof image === "string" && (image.startsWith("data"))){

                        let collectionInfo = await fieldsService.readOne(collection_name, file_field_name);
                        console.log(collectionInfo);
                        let folder = collectionInfo?.meta?.options?.folder;
                        if(!!folder){
                                body.folder = folder;
                        }

                        console.log("DecodeBase64");
                        let decodedObj = decodeBase64Image(image);
                        console.log("Upload image");
                        body.type = decodedObj.type;
                        body.filename_download = "test"+"."+decodedObj.extension;
                        let imageBuffer = decodedObj.data;
                        try{
                        console.log(body);
                        const filename = await filesService.uploadOne(imageBuffer, body);
                        console.log("filename");
                        console.log(filename);
                        payload.image = filename;
                        } catch(err){
                                console.log(err);
                        }
                }
        }
        payload.image_upload_base64 = payload.image;
        console.log(payload);
        return payload;
}

```

**Reply by NilsBaumgartner1994** [reply]
Created for Directus 9.8

**Reply by ejabu** [reply]
is there any Directus extension for this ? how can we upload image by base64 in directus ?

**Reply by NilsBaumgartner1994** [reply]
Step 1. install 'directus-cut'

Step 2. add a hook file in directus for extample: /extensions/hooks/base64hook/index.js

index.js
```
const directusCut = require("directus-cut")


const beforeHook = null;
const afterHook = null;
const collection_name = "<YOUR_COLLECTION>";
const file_field_name = "<YOUR_FILE_FIELD_IN_THE_COLLECTION>";
module.exports = (registerFunctions, context) => {
  registerFunctions.filter(
      collection_name + '.items.update',
      async (payload: any, input: any, {database, schema, accountability}) => {
          directusCut.BackendUtils.BackendExtensions.BackendHooks.ImageBase64UploadHook.registerHook(beforeHook, afterHook, collection_name, file_field_name, payload, input, database, schema, accountability, registerFunctions, context);
  }
}
ImageBase64UploadHook

```

**Comment by NilsBaumgartner1994** [comment]
I can post my solution here later. I just added a small hook, which converts images from base64. It may only work with local storage and not aws.Beste Grüße,Nils BaumgartnerAm 22.11.2022 um 02:46 schrieb Ejabu ***@***.***>:﻿
is there any Directus extension for this ? how can we upload image by base64 in directus ?

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
